src/main/java/org/opensearch/neuralsearch/processor/NeuralSparseTwoPhaseProcessor.java

```diff
@@ -0,0 +1,246 @@
+        }
+        return Map.of(UP_THRESHOLD, highScoreTokensSetOnce, DOWN_THRESHOLD, lowScoreTokensSetOnce);
+    }
+
+    /**
+     * Factory to create NeuralSparseTwoPhaseProcessor, provide default parameter,
+     *
+     */
+    public static class Factory implements Processor.Factory<SearchRequestProcessor> {
+        @Override
+        public NeuralSparseTwoPhaseProcessor create(
+            Map<String, Processor.Factory<SearchRequestProcessor>> processorFactories,
+            String tag,
+            String description,
+            boolean ignoreFailure,
+            Map<String, Object> config,
+            PipelineContext pipelineContext
+        ) throws IllegalArgumentException {
+
+            boolean enabled = ConfigurationUtils.readBooleanProperty(TYPE, tag, config, ENABLE_KEY, DEFAULT_ENABLED);
+            Map<String, Object> twoPhaseConfigMap = ConfigurationUtils.readOptionalMap(TYPE, tag, config, PARAMETER_KEY);
+
+            float ratio = DEFAULT_RATIO;
+            float window_expansion = DEFAULT_WINDOW_EXPANSION;
+            int max_window_size = DEFAULT_MAX_WINDOW_SIZE;
+            if (Objects.nonNull(twoPhaseConfigMap)) {
+                ratio = ((Number) twoPhaseConfigMap.getOrDefault(RATIO_KEY, ratio)).floatValue();
+                window_expansion = ((Number) twoPhaseConfigMap.getOrDefault(EXPANSION_KEY, window_expansion)).floatValue();
+                max_window_size = ((Number) twoPhaseConfigMap.getOrDefault(MAX_WINDOW_SIZE_KEY, max_window_size)).intValue();
+            }
+
+            return new NeuralSparseTwoPhaseProcessor(tag, description, ignoreFailure, enabled, ratio, window_expansion, max_window_size);
+        }
+    }
+
+    private QueryBuilder getNestedQueryBuilderFromNeuralSparseQueryBuilderMap(
+        final Multimap<NeuralSparseQueryBuilder, Float> queryBuilderFloatMap
+    ) {
+        BoolQueryBuilder boolQueryBuilder = new BoolQueryBuilder();
+        queryBuilderFloatMap.asMap().forEach((neuralSparseQueryBuilder, boosts) -> {
+            float reduceBoost = boosts.stream().reduce(0.0f, Float::sum);
+            final float finalBoost = neuralSparseQueryBuilder.boost() * reduceBoost;
+            boolQueryBuilder.should(neuralSparseQueryBuilder.boost(finalBoost));
+        });
+        return boolQueryBuilder;
+    }
+
+    private float getOriginQueryWeightAfterRescore(final SearchSourceBuilder searchSourceBuilder) {
+        if (Objects.isNull(searchSourceBuilder.rescores())) {
+            return 1.0f;
+        }
+        return searchSourceBuilder.rescores()
+            .stream()
+            .map(rescorerBuilder -> ((QueryRescorerBuilder) rescorerBuilder).getQueryWeight())
+            .reduce(1.0f, (a, b) -> a * b);
+    }
+
+    private Multimap<NeuralSparseQueryBuilder, Float> collectNeuralSparseQueryBuilder(final QueryBuilder queryBuilder, float baseBoost) {
+        Multimap<NeuralSparseQueryBuilder, Float> result = ArrayListMultimap.create();
```

Comment:
Why we use Multimap here? Do we have multiple values for one field?

Reply:
If there has two same neuralsparse querybuilder in a compound query, will have same key.